rust-runtime/aws-smithy-runtime/src/client/retries/classifier.rs

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use aws_smithy_http::result::SdkError;
+use aws_smithy_runtime_api::client::retries::RetryReason;
+use aws_smithy_types::retry::{ErrorKind, ProvideErrorKind};
+
+/// A retry classifier for checking if an error is modeled as retryable.
+#[derive(Debug)]
+pub struct ModeledAsRetryableClassifier;
+
+impl ModeledAsRetryableClassifier {
+    /// Check if an error is modeled as retryable, returning a [`RetryReason::Error`] if it is.
+    pub fn classify_error<E: ProvideErrorKind, R>(
+        &self,
+        error: &SdkError<E, R>,
+    ) -> Option<RetryReason> {
+        match error {
+            SdkError::ServiceError(inner) => inner
+                .err()
+                .retryable_error_kind()
+                .map(|err| RetryReason::Error(err)),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SmithyErrorClassifier;
+
+impl SmithyErrorClassifier {
+    pub fn classify_error<E, R>(&self, result: &SdkError<E, R>) -> Option<RetryReason> {
+        match result {
+            SdkError::TimeoutError(_err) => Some(RetryReason::Error(ErrorKind::TransientError)),
+            SdkError::ResponseError { .. } => Some(RetryReason::Error(ErrorKind::TransientError)),
+            SdkError::DispatchFailure(err) if (err.is_timeout() || err.is_io()) => {
+                Some(RetryReason::Error(ErrorKind::TransientError))
+            }
+            SdkError::DispatchFailure(err) => err.is_other().map(|err| RetryReason::Error(err)),
+            _ => None,
+        }
+    }
+}
+
+const TRANSIENT_ERROR_STATUS_CODES: &[u16] = &[500, 502, 503, 504];
+
+/// A retry classifier that will treat HTTP response with those status codes as retryable.
+/// The `Default` version will retry 500, 502, 503, and 504 errors.
+#[derive(Debug)]
+pub struct HttpStatusCodeClassifier {
+    retryable_status_codes: Vec<u16>,
```

Comment:
This could be more memory efficient by making it a `Cow<'static, [u16]>` since the majority of the time it will be a constant.